src/metabase/actions.clj

```diff
@@ -136,6 +114,29 @@
           (swap! *misc-value-cache* assoc unique-key value))
         value)))
 
+(defn do-when-actions-enabled
+  "Implementation for [[when-actions-enabled]]."
+  [{db-settings :settings db-id :id driver :engine db-name :name :as db} thunk]
+  ;; make sure the Driver supports Actions.
+  (when-not (driver/database-supports? driver :actions db)
+    (throw (ex-info (i18n/tru "{0} Database {1} does not support actions."
+                              (u/qualified-name driver)
+                              (format "%d %s" db-id (pr-str db-name)))
+                    {:status-code 400, :database-id db-id})))
+  ;; bind Database-local Settings for this Database and the misc value cache
+  (binding [setting/*database-local-values* db-settings
+            *misc-value-cache*              (atom {})]
+    ;; make sure Actions are enabled for this Database
+    (when-not (database-enable-actions)
+      (throw (ex-info (i18n/tru "Actions are not enabled.")
+                      {:status-code 400, :database-id db-id})))
+    (thunk)))
+
+(defmacro when-actions-enabled
+  "Execute the body if actions are enabled for the given database, otherwise throw."
+  [database & body]
+  `(do-when-actions-enabled ~database (fn [] ~@body)))
```

Comment:
The name and docstrings are misleading here. To understand what this does I have to read the whole thing. It does a lot of things outside "Execute the body if actions are enabled for the given database, otherwise throw". We should be able to separate the part that checks whether actions are enabled from the parts that do other things like the parts binds *misc-value-cache*. It also looks like *misc-value-cache* wasn't needed anywhere other than for `perform-action!`, so it shouldn't be bound here. 

I think once you've separated these parts you shouldn't need a macro for checking whether actions are enabled. If you'd rather not duplicate error handling code, it could just be a call to a function like `(check-database-enable-actions database)`.

Also, I would question whether we need to check whether the driver supports actions here, since we are already checking that the database has actions enabled. How is it possible that a driver doesn't support actions, yet actions are enabled for the database? If it's impossible, then this test can be removed.